### PR TITLE
Fix `BinaryDirichletFFN` and  `MulticlassDirichletFFN` predictors

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
As identified in previous PR (https://github.com/chemprop/chemprop/pull/959#discussion_r1747656884), the predictors for `BinaryDirichletFFN` are not correct. Based on the original [paper](https://arxiv.org/abs/1806.01768), the logits should be passed into an activation function like `ReLU` to get non-negative output ($\alpha_i$). Here, we use `SoftPlus`. Then, the corresponding Dirichlet distribution has parameters:

$$ \alpha_i = f(\mathbf{x}_i \mid \Theta) + 1 $$

where $\Theta$ is the trained network. The probability of each class can be calculated as:

$$ \frac{\alpha_i}{\mathbf{S}} $$

where $\mathbf{S} = \sum_{i=1}^{K} \alpha_i$ and  $K$ is the number of class labels. The uncertainty can be calculated  as:

$$ \frac{K}{\mathbf{S}} $$

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

File: chemprop/nn/predictors.py [hunk 1]
```diff
@@ -294,18 +297,16 @@ class MulticlassDirichletFFN(MulticlassClassificationFFN):
     _T_default_metric = CrossEntropyMetric
 
     def forward(self, Z: Tensor) -> Tensor:
-        Y = super().forward(Z).reshape(len(Z), -1, self.n_classes)
+        Y = super().train_step(Z)
 
-        Y = Y.softmax(-1)
-        Y = F.softplus(Y) + 1
+        alpha = F.softplus(Y) + 1
```
Comment (shihchengli) [inline chemprop/nn/predictors.py:303]: The uncertainty can be calculated as `u = self.n_classes / alpha.sum(-1)`. However, since the contract we signed specifies 1 target, I only return the predicted property. Additionally, as we discussed here (https://github.com/chemprop/chemprop/pull/959#issuecomment-2261639212), the multiclass classification output should be in the shape of b x t x c.

File: chemprop/nn/predictors.py [hunk 2]
```diff
@@ -229,14 +229,17 @@ class BinaryDirichletFFN(BinaryClassificationFFNBase):
...
-        Y = beta / (alpha + beta)
+        Y = super().forward(Z).reshape(len(Z), -1, 2)
+
+        alpha = F.softplus(Y) + 1
 
-        return Y.reshape(Y.shape[0], -1, 1)
+        u = 2 / alpha.sum(-1)
+        Y = alpha / alpha.sum(-1, keepdim=True)
+
+        return torch.stack((Y[..., 1], u), dim=2)
```
Comment (shihchengli) [inline chemprop/nn/predictors.py:239]: Using `[..., 1]` to get the predicted probability for class 1.

File: tests/cli/test_cli_classification_mol_multiclass.py [hunk 3]
```diff
@@ -29,14 +29,21 @@ def test_train_quick(monkeypatch, data_path):
         "3",
         "--num-workers",
         "0",
-        "--task-type",
-        "multiclass",
         "--show-individual-scores",
+        "--accelerator",
+        "cpu",
```
Comment (shihchengli) [inline tests/cli/test_cli_classification_mol_multiclass.py:34]:
```suggestion
```

**Comment by hwpang** [comment]
@shihchengli Thanks for the PR! Could you add integration test for these two predictors to make sure they indeed fit correctly?

**Comment by shihchengli** [comment]
@hwpang I have added the integration test. One of the tests failed because the reported accuracy for the overfitting setting with `AtomMessagePassing` is 0.9895, which is lower than 0.99. Do you have any suggestions on how to fix this?

**Comment by hwpang** [comment]
> @hwpang I have added the integration test. One of the tests failed because the reported accuracy for the overfitting setting with `AtomMessagePassing` is 0.9895, which is lower than 0.99. Do you have any suggestions on how to fix this?

Does this accuracy improve if you use either more epochs or a deeper network? I think all we want to show is that it can overfit the small set. @davidegraff What's your though on this?

**Comment by shihchengli** [comment]
@hwpang I tested it on Linux. With 100 epochs, the AUC-ROC is already 0.9984. If 500 epochs is used, the AUC-ROC is 1. Using more epochs might be able to solve the problem.

**Comment by davidegraff** [comment]
It’s definitely an epochs problem. Adding more depth will likely make the problem more challenging—happy to explain the last point if it’s not clear.

**Comment by hwpang** [comment]
> It’s definitely an epochs problem. Adding more depth will likely make the problem more challenging—happy to explain the last point if it’s not clear.

Yeah true. More epochs can force the network to overfit to the small set of training data. I was thinking a deeper network is more expressive and thus can better overfit the data, but will likely need more epochs to converge and thus will make the training accuracy worse given the same epochs.

**Comment by davidegraff** [comment]
I'd be careful not to confuse depth and capacity for GNNs. You're not simply increasing the receptive field and adding more non-linearities; it also leads to oversmoothing[^1], which is likely going to do more harm than good.

[^1]: https://arxiv.org/abs/2303.10993

**Review by KnathanM — APPROVED** [review]
To my eye, this matches the v1 code so v1 dirichlet models can probably be converted to v2 models. But I think that would be rare for someone to want to do, so I won't ask you to add check that with the v1_to_v2 convert script.
